add support for asynchronous building and pushing of profiles

PhilTaken opened the pull request:
NB: I have opened this Pull Request as a *draft* since I intend to continue working on it by improving the logging output.
Since the builds are started concurrently, the build progress information is pretty much useless and flickers a lot.

# Problem

The current implementation builds every single profile synchronously, including remote builds. 
Since remote builds were introduced back in #175, remote builds could be pipelined to deploy
new configurations in a more time-efficient manner.

# Solution

Build configurations that support remote builds concurrently.

---

Sidenote: I have decided to continue building local builds in a synchronous manner because I have run into hardware deadlocks previously when trying to evaluate and/or build multiple systems at the same time.

I have tested this code by deploying to my personal infrastructure (https://github.com/philtaken/dotfiles) and it works as intended.

rvem commented on src/cli.rs:
```diff
@@ -585,13 +587,33 @@ async fn run_deploy(
+    // await both the remote builds and the local builds to speed up deployment times
+    try_join!(
+        // remote builds can be run asynchronously since they do not affect the local machine
+        try_join_all(remote_builds.into_iter().map(|data| async {
```
From `try_join_all` docs:
> If any future returns an error then all other futures will be canceled and an error will be returned immediately

IMO, it might be better to wait for all futures to be completed/failed instead of canceling everything on the first failure, thus the non-failed builds will complete despite the error in an unrelated build. What do you think?

rvem replied:
> Since the builds are started concurrently, the build progress information is pretty much useless and flickers a lot.

I'm not sure how viable it is to implement, but perhaps, it's possible to collect `stdout` and `stderr` for each future, if it is, we can collect and report detailed output synchronously once all remote builds are completed

PhilTaken replied:
> IMO, it might be better to wait for all futures to be completed/failed instead of canceling everything on the first failure, thus the non-failed builds will complete despite the error in an unrelated build. What do you think?

I'll see how easy it would be to implement that, if it is that might be worth checking out so that consecutive builds don't have to build as much.

> perhaps, it's possible to collect stdout and stderr for each future, if it is, we can collect and report detailed output synchronously once all remote builds are completed

I was considering that maybe one line from the bottom per build would be neat e.g.:

[previous output]
host 1: [build progress] 
host 2: [build progress]

This will, however, require some more fine grained control over the terminal output i.e. raw output instead of cooked.

rvem replied:
> would be neat

Agree, but indeed sounds quite non-trivial

rvem replied:
Oh, another somewhat conceptual concern. Profiles can potentially depend on each other, so perhaps it's worth doing the parallelization on per-host basis instead of pre-profile

PhilTaken commented:
mentioning #46 for visibility

the current version works as expected, the only issue is the log flickering with multiple invocations of nix writing to stdout in parallel. I was considering utilising the raw-internal-json output similarly to how nix-output-monitor does it but that might be out of scope here 🤔

PhilTaken commented:
activation is fully synchronous but that is usually the part that takes the least amount of time

rvem commented on src/cli.rs:
```diff
@@ -585,12 +588,43 @@ async fn run_deploy(
+        async {
+            // run local builds synchronously to prevent hardware deadlocks
+            for data in &local_builds {
+                deploy::push::build_profile(data).await.unwrap();
```
I'm not a huge fan of using "partial functions", isn't unhandled panic from `unwrap` going to kill the main thread?
Also, AFAICS, at the moment we completely ignore non-remote build/push results

rvem commented:
The new approach seems good 👍

PhilTaken commented:
there are a still lot of unwraps that need to be handled better, this pr is still far from finished but progress has been made:

I added some progress indicators using the indicatif crate for each concurrent build 🥳 

This is what they look like (image the spinner spinning and the text changing as nix builds stuff:

![image (1)](https://github.com/user-attachments/assets/a5a74b5c-87eb-4ea4-a6f5-0c0baa9a65cd)

For remote and builds each host gets its own progress spinner that is reused for each profile since they are built in order anwyas.

I changed some data structures around to make working with them in async contexts easier.
The nested references are somewhat ok until you have to make them work in a async context where you need send objects. deep cloning by hand really isn't a great solution and it really doesnt make any difference performance-wise to add some clones here and there. 
Nobody is going to deploy to >10k hosts and if they would, cloning the deploy data a few times would be the least of their problems. 
This also remove a bunch of generic lifetime specifier which I am not sad about.

PhilTaken commented:
With these two last commits I would declare this PR somewhat usable 🥳 

If anyone wants to try it out with their own setups I would be very happy to hear how it works for you, I am always open for feedback of any kind

ManoftheSea commented:
Hi all,
I tried this out, with my setup in ManoftheSea/SeaofDirac/trunk, on targets [littlecreek, crunchbits, technetium].  I saw that it ran the checks for all the systems, which meant building them locally, then proceeded with pushing profiles to the systems in parallel (which resulted in building them remotely, I believe).  After all three were ready, they deployed in sequence, and as it was just an update, they all completed happily.

This works well for these three systems, which are a pair of VPSes and a local (to me) server; were I trying to deploy all the systems in my home network, I would prefer instead that my development machine (personal laptop) be able to push the profiles to a single server (technetium), so that it can then act as a substituter for e.g. aluminium and nickel.  Maybe provide a "build-host" attribute per-host, with a default of the hostself?  e.g. deploy.nickel.buildHost = "technetium"; deploy.littlecreek.buildHost = "littlecreek";

ChecksumDev commented:
Hi there, our team at [Pyro](https://pyro.host) utilized this PR and observed a massive speedup deploying to a our node cluster and for our initial deployments we see a huge speed increase of at minimum 2x, however there might be a problem. 

I'm deploying from Ubuntu WSL, so that might be the culprit but it seems after every deployment in the same terminal session only half will be working in parallel, and then half again, until its back to single threaded (even though the spinners are still there). It might not be halfing but there is at minimum diminishing returns over multiple command invocations.

Here's a successful one if you're curious, I haven't been able to record the above yet though.
I figured with such a big cluster our network would likely be prime testing ground for this PR.

I'm really excited for this to get merged and I'll test anything necessary to speed things along.

https://asciinema.org/a/dirahL9YYWFaT0lORkU98nX8s

P.s. A huge potential speedup would be processing checks in a parallel fashion as well, as that takes up a huge chunk of time, though I'm not sure how viable this is.

PhilTaken commented:
> Hi there, our team at [Pyro](https://pyro.host/) utilized this PR and observed a massive speedup deploying to a our node cluster and for our initial deployments we see a huge speed increase of at minimum 2x

oh that is great to hear! thanks for the feedback.

> it seems after every deployment in the same terminal session only half will be working in parallel, and then half again, until its back to single threaded (even though the spinners are still there). It might not be halfing but there is at minimum diminishing returns over multiple command invocations.

hmm, I'll have to investigate how that might happen. I'll add some more debug output for each thread so it'll easier to debug on your side. This pr pretty much just spawns a bunch of threads using high-level abstractions so I'm curious how that might happen 🤔 

 > P.s. A huge potential speedup would be processing checks in a parallel fashion as well, as that takes up a huge chunk of time, though I'm not sure how viable this is.

unfortunately, the checks are limited by nix's eval speed. there is some work going on on detsys's side (see e.g. https://determinate.systems/posts/parallel-nix-eval/) but afaiu nix flake checks are single-threaded in nature. 

you could attempt to get around that by building all systems in parallel batches using `gnu parallel` although I am not sure how feasible that is. I do not really have access to hardware that could build even two systems in parallel...

ChecksumDev reviewed (approved):
Since my previous comment, we've continued using this pull request, now with approximately 30 nodes, and have demonstrated no further problems. To us, at least, it seems to have identical parity to its single-threaded counterpart on `main`.

Performance might be improvable and the visual aspects could be worked on, but so far nothing has broken down and it's been smooth sailing, so I'd call this pretty stable! 👀

rvem commented on src/push.rs:
```diff
@@ -169,48 +174,81 @@ pub async fn build_profile_remotely(data: &PushProfileData<'_>, derivation_name:
+
+        let mut child = copy_command.stderr(Stdio::piped()).stdout(Stdio::piped()).spawn().expect("failed to spawn nix copy command");
+
+        if let Some(pb) = &data.deploy_data.progressbar {
```
Is there a chance to abstract away common parts that handle subprocess outputs?
Handling for `nix copy` and `nix build` looks completely identical and it might be easy to update only one of them when needed and forget about the second

rvem reviewed (commented):
It took me waaay too much time to get back to this PR, I'm sorry about that:(
I tested these changes, UX looks neat, code changes themselves look good too, though I have a few minor concerns/suggestions

PhilTaken commented:
just a small update for everybody interested in this: I am still working on it, didnt have all too much time recently. This weekend i've got some more time so I'll address all of rvem's points above 🫡

ChecksumDev commented:
Thank you for the update! Our team was considering taking over this PR and aiding its development, but its good to know there's still activity. We use this branch religiously and are free to test any new functionality / bug fixes.

![image](https://github.com/user-attachments/assets/25cdfdaa-9d17-41a7-aeca-02fbf7de2aa7)

PhilTaken commented on src/cli.rs:
```diff
@@ -577,38 +583,151 @@ async fn run_deploy(
+                    Err(ref e) => {
+                        // TODO set style to red X
+                        pb.finish_with_message(format!("Error: {}", e.to_string()));
+                    }
```
good point, just requires a faux future

PhilTaken commented:
> We use this branch religiously and are free to test any new functionality / bug fixes.

that is a huge number of hosts, thanks for a large scale test on this!

did you encounter the issue with the deployments not running in parallel again? I could not find anything in the code that could lead to such an issue

rvem commented on src/push.rs:
```diff
@@ -287,7 +325,7 @@ pub async fn build_profile(data: PushProfileData<'_>) -> Result<(), PushProfileE
 }
 
-pub async fn push_profile(data: PushProfileData<'_>) -> Result<(), PushProfileError> {
+pub async fn push_profile(data: &PushProfileData) -> Result<(), PushProfileError> {
```
I see, thanks for the clarification

Sereja313 committed:
feat: async build-and-push pipeline with indicatif progress indicators

- build and push profiles asynchronously using tokio tasks
- group remote builds by host for cleaner output
- continue building when a remote build fails and report errors afterwards
- wrap logger with a custom implementation
- add indicatif progress indicators for remote and local builds
- improve error handling for async deployments
- address PR feedback
- add checkmarks (success) / red cross (error) indicators for finished/aborted deployments

Co-authored-by: Philipp Herzog <philipp.herzog@protonmail.com>

Sereja313 committed:
chore: improve progress information
